usvm-jvm/src/main/kotlin/org/usvm/api/util/JcTestStateResolver.kt

```diff
@@ -286,25 +335,30 @@ abstract class JcTestStateResolver<T>(
         val classTypeField = ctx.classTypeSyntheticField
         val classTypeLValue = UFieldLValue(ctx.addressSort, ref, classTypeField)
 
-        val classTypeRef = memory.read(classTypeLValue) as? UConcreteHeapRef
+        val memoryToResolveClassType = if (isStaticHeapRef(ref)) finalStateMemory else model
+
+        val classTypeRef = memoryToResolveClassType.read(classTypeLValue) as? UConcreteHeapRef
             ?: error("No type for allocated class")
 
-        val classType = memory.typeStreamOf(classTypeRef).first()
+        val classType = memoryToResolveClassType.typeStreamOf(classTypeRef).first()
 
         return decoderApi.createClassConst(classType)
     }
 
     abstract fun allocateString(value: T): T
 
     fun resolveAllocatedString(ref: UConcreteHeapRef): T {
-        val stringConstant = stringConstants.entries.singleOrNull { it.value == ref }
-        if (stringConstant != null) {
-            return decoderApi.createStringConst(stringConstant.key)
-        }
-
         val valueField = ctx.stringValueField
         val strValueLValue = UFieldLValue(ctx.typeToSort(valueField.fieldType), ref, valueField.field)
-        val strValue = resolveLValue(strValueLValue, valueField.fieldType)
+
+        val strValue = if (isStaticHeapRef(ref)) {
+            withMode(ResolveMode.CURRENT) {
```

Comment:
Why we use `withMode` here instaed of `finalStateMemory`?

Reply:
Because sometimes we resolve strings in the model. If we read here from the final memory, only the address will be correct, but the content of the string will be constructed from the model. That led us to an inconsistent state